drivers/ethernet/eth_lowRISC.c

```diff
@@ -0,0 +1,357 @@
+
+/*
+ * Copyright (c) Recogni.
+ *
+ * LowRISC Ethernet driver
+ */
+
+
+#define DT_DRV_COMPAT lowrisc_eth
+
+#define LOG_MODULE_NAME eth_lowRISC
+#define LOG_LEVEL CONFIG_ETHERNET_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
+#include <sys/types.h>
+#include <zephyr.h>
+#include <device.h>
+#include <string.h>
+#include <errno.h>
+#include <drivers/gpio.h>
+#include <drivers/spi.h>
+#include <net/net_pkt.h>
+#include <net/net_if.h>
+#include <net/ethernet.h>
+#include <ethernet/eth_stats.h>
+#include <assert.h>
+#include "eth_lowRISC.h"
+#include "mdiobb.h"
+#include "rtl8211_phy.h"
+
+
+#if defined(CONFIG_ETH_LOWRISC_VERBOSE_DEBUG)
+#define hexdump(_buf, _len, fmt, args...)				\
+({									\
+	const size_t STR_SIZE = 80;					\
+	char _str[STR_SIZE];						\
+									\
+	snprintk(_str, STR_SIZE, "%s: " fmt, __func__, ## args);	\
+									\
+	LOG_HEXDUMP_DBG(_buf, _len, log_strdup(_str));			\
+})
+#else
+#define hexdump(args...)
+#endif
+
+static void inline eth_copyout(struct net_local_lr *priv, u8 *data, int len)
+{
+    int i, rnd = ((len - 1) | 7) + 1;
+    volatile u64 *eth_base = (volatile u64 *)(priv->ioaddr);
+    if (!(((size_t)data) & 7))
+    {
+        u64 *ptr = (u64 *)data;
+        for (i = 0; i < rnd / 8; i++) eth_base[TXBUFF_OFFSET / 8 + i] = ptr[i];
+    }
+    else  // We can't unfortunately rely on the skb being word aligned
+    {
+        u64 notptr;
+        for (i = 0; i < rnd / 8; i++)
+        {
+            memcpy(&notptr, data + (i << 3), sizeof(u64));
+            eth_base[TXBUFF_OFFSET / 8 + i] = notptr;
+        }
+    }
+}
+
+
+static int lr_send(const struct device *dev, struct net_pkt *pkt)
```

Comment:
We should group the `eth_` and `lr_` functions together. Especially copyin/copyout.